app/controllers/payments_controller.rb

```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
+                               payment_processor: Payment::PAYMENT_PROCESSOR_KLARNA,
+                               start_date: start_date,
+                               expire_date: expire_date)
+
+    payment_data = { id: @payment.id,
+                     user_id: user_id,
+                     type: payment_type,
+                     currency: 'SEK',
+                     item_price: item_price,
+                     paid_item: paid_item,
+                     urls:  klarna_order_urls(@payment.id, user_id,
+                                              company_id, payment_type) }
+
+    # Invoke Klarna API - returns an order to be used for checkout
+    klarna_order = KlarnaService.create_order(payment_data)
+
+    # Save payment and render checkout form
+    klarna_id = klarna_order['order_id']
+    @payment.klarna_id = klarna_id
+    @payment.status = Payment.order_to_payment_status(klarna_order['status'])
     @payment.save!
 
-  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid  => exc
-    @payment.destroy if @payment.persisted?
+    log_klarna_activity('Create Order', 'info', @payment.id, klarna_id)
 
-    log_hips_activity('create order', 'error', nil, @hips_id, exc)
+    @html_snippet = klarna_order['html_snippet']
 
-    log_hips_activity('create order', 'error', nil, @hips_id, exc.cause)
+  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid  => exc
+    @payment.destroy if @payment&.persisted?
+
+    log_klarna_activity('create order', 'error', nil, klarna_id, exc)
 
     helpers.flash_message(:alert, t('.something_wrong',
                                     admin_email: ENV['SHF_REPLY_TO_EMAIL']))
 
+    notify_slack_of_exception(exc, __method__)
+
+    redirect_back fallback_location: root_path
+  end
+
+  def confirmation
+    # https://developers.klarna.com/documentation/klarna-checkout/in-depth/confirm-purchase
+
+    payment_id = params[:id]
+    klarna_id = params[:klarna_id]
+    raise 'No Klarna order ID' unless klarna_id
+
+    klarna_order = handle_order_confirmation(klarna_id, payment_id)
+
+    account_page_link = helpers.link_to(t('menus.nav.users.your_account').downcase,
+                                        user_path(params[:user_id]))
+
+    helpers.flash_message(:notice, t('.success_html',
+                                     account_page_link: account_page_link))
+
+    log_klarna_activity('Order Confirmation', 'info', payment_id, klarna_id)
+
+    @html_snippet = klarna_order['html_snippet']
+
+  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid => exc
+    log_klarna_activity('Order Confirmation', 'error', payment_id, klarna_id, exc)
+
+    notify_slack_of_exception(exc, __method__)
+
+    helpers.flash_message(:alert, t('payments.create.something_wrong',
+                                    admin_email: ENV['SHF_REPLY_TO_EMAIL']))
+
     redirect_back fallback_location: root_path
   end
 
+  def klarna_push
```

Comment:
I don't see a lot of duplication (?)

Reply:
Your rescue clauses are duplicates.  At the least you should have a method that returns the list of errors you're checking/rescuing and then use that instead of always listing the same classes.
You are also always calling `log_klarna_activity` and `notify_slack_of_exception`.  Create 1 method that calls them both and then just use that 1 method.

In general the methods have very similar designs (overall pattern). That is an indicator that there may be common behavior that can be refactored into common methods that can be called and/or that a more general 'outer' method can be used with a `yield` statement to handle  behavior specific to each common original method.